Make some utilities public...

davidfestal opened the pull request:
Make some CRD utilities public for them to be used by the KCP `dynamic` virtual workspaces.

Signed-off-by: David Festal <dfestal@redhat.com>

sttts commented:
Lgtm

ncdc commented:
Do we intend to get this upstreamed?

davidfestal commented:
> Do we intend to get this upstreamed?

What's your thought on this @sttts ?

ncdc commented:
I would think we'd want to

sttts commented:
> Do we intend to get this upstreamed?

Likely. But the value is limited. These "make x public" commits are not that hard to carry.

sttts commented:
Btw, it might be better to add some helpers instead of just making things public:
```golang
type X = x
func Foo(...) .... { return foo(...) }
```
and put those into a `patch_foo.go` file. Less conflicts.

davidfestal commented:
> Btw, it might be better to add some helpers instead of just making things public:
> 
> ```go
> type X = x
> func Foo(...) .... { return foo(...) }
> ```
> 
> and put those into a `patch_foo.go` file.

Good idea. I'll try that.

ncdc commented:
I'd also suggest we be diligent with our PR titles and commits. e.g. update this title to be more descriptive about what this is for. Also, we need to be more consistent with our commit message prefixes - UPSTREAM vs. UPSTREAM(carry) vs ... I've been meaning to document these in kcp, but haven't gotten to it yet.

davidfestal committed:
UPSTREAM: <carry>: Make the function public ...

... to use it on the KCP side
in the `virtual/dynamic/apiserver` virtual workspace package.

Signed-off-by: David Festal <dfestal@redhat.com>

davidfestal committed:
UPSTREAM: <carry>: Add public wrapper and factory functions ...

... to make the `scopeWithFieldManager` function
as well as the `unstructuredNegotiatedSerializer`
and `unstructuredDefaulter` structs available
from external packages.

This was initially required by KCP
in the `virtual/dynamic/apiserver` virtual workspace package.

Signed-off-by: David Festal <dfestal@redhat.com>

davidfestal commented:
@sttts @ncdc Can you confirm that it's now what you were expecting ?

ncdc commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/patch_customresource_handler.go:
```diff
@@ -0,0 +1,28 @@
+/*
```
I see this file is part of an UPSTREAM (non carry) commit. If we intend to upstream this, the file should have a better name. I know @sttts suggested doing it this way - we need to resolve how we want to handle this.

davidfestal replied:
Maybe in this case I simply rename the function without the new file. @sttts what's your preferred way in this specific simple case ?

sttts replied:
Would use the patch file. Value of upstreaming is low, and chance is that ppl don't like leaking internals. Not a war worth to fight for now IMO.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1230,6 +1248,17 @@ func (c UnstructuredCreator) New(kind schema.GroupVersionKind) (runtime.Object,
+		structuralSchemas: structuralSchemas,
+		structuralSchemaGK: structuralSchemaGK,
+	}
+}
```
move to patch file

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1124,6 +1124,24 @@ func defaultDeprecationWarning(deprecatedVersion string, crd apiextensionsv1.Cus
+		preserveUnknownFields: preserveUnknownFields,
+	}
+}
+
```
move to patch file

davidfestal commented:
@sttts @ncdc Everything new is in the patch file now, as `<carry>` commits. Is it OK for you to approve ?

ncdc commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/patch_customresource_handler.go:
```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2022?

davidfestal replied:
Fixed

sttts commented:
/lgtm
